crates/sui-indexer-alt-graphql/src/api/types/event/mod.rs

```diff
@@ -77,3 +108,125 @@ impl Event {
         ))
     }
 }
+
+impl Event {
+    /// Paginates events based on the provided filters and page parameters.
+    pub(crate) async fn paginate(
+        ctx: &Context<'_>,
+        scope: Scope,
+        page: Page<CEvent>,
+        filter: filter::EventFilter,
+    ) -> Result<Connection<String, Event>, RpcError> {
+        let mut c = Connection::new(false, false);
+
+        let pg_reader: &PgReader = ctx.data()?;
+        let watermarks: &Arc<Watermarks> = ctx.data()?;
+
+        // TODO: (henry) Use watermarks once we have a strategy for kv pruning.
+        let reader_lo = 0;
+        let global_tx_hi = watermarks.high_watermark().transaction();
+
+        let cp_bounds = checkpoint_bounds(
+            filter.after_checkpoint.map(u64::from),
+            filter.at_checkpoint.map(u64::from),
+            filter.before_checkpoint.map(u64::from),
+            reader_lo,
+            scope.checkpoint_viewed_at(),
+        )
+        .context("Failed to calculate checkpoint bounds")?;
+
+        let tx_bounds = tx_bounds(ctx, &cp_bounds, global_tx_hi).await?;
+        let pg_tx_bounds = pg_tx_bounds(&page, tx_bounds);
+
+        // TODO: (henry) update query to select from ev_emit_mod or ev_struct_inst based on filters.
+        let query = query!(
```

Comment:
Ah interesting, for the unfiltered case, I didn't think about the fact that we don't want to fetch transactions that have no events!